src/libpmempool/replica.c

```diff
@@ -528,29 +658,469 @@ check_checksums_and_signatures(struct pool_set *set,
 			if (!util_checksum(hdr, sizeof(*hdr), &hdr->checksum, 0,
 					POOL_HDR_CSUM_END_OFF)) {
 				ERR("invalid checksum of pool header");
-				rep_hs->part[p] |= IS_BROKEN;
+				rep_hs->part[p].flags |= IS_BROKEN;
 			} else if (util_is_zeroed(hdr, sizeof(*hdr))) {
-					rep_hs->part[p] |= IS_BROKEN;
+					rep_hs->part[p].flags |= IS_BROKEN;
 			}
 
 			enum pool_type type = pool_hdr_get_type(hdr);
 			if (type == POOL_TYPE_UNKNOWN) {
 				ERR("invalid signature");
-				rep_hs->part[p] |= IS_BROKEN;
+				rep_hs->part[p].flags |= IS_BROKEN;
+			}
+		}
+	}
+	return 0;
+}
+
+/*
+ * replica_badblocks_recovery_file_save -- save bad blocks in the bad blocks
+ *                                         recovery file before clearing them
+ */
+static int
+replica_badblocks_recovery_file_save(struct part_health_status *part_hs)
+{
+	LOG(3, "part_health_status %p", part_hs);
+
+	ASSERTeq(part_hs->recovery_file_exists, 1);
+	ASSERTne(part_hs->recovery_file_name, NULL);
+
+	struct badblocks *bbs = &part_hs->bbs;
+	char *path = part_hs->recovery_file_name;
+	int ret = -1;
+
+	int fd = os_open(path, O_WRONLY | O_TRUNC);
+	if (fd < 0) {
+		ERR("!opening bad block recovery file failed -- '%s'", path);
+		return -1;
+	}
+
+	FILE *recovery_file_name = os_fdopen(fd, "w");
+	if (recovery_file_name == NULL) {
+		ERR(
+			"!opening a file stream for bad block recovery file failed -- '%s'",
+			path);
+		os_close(fd);
+		return -1;
+	}
+
+	/* save bad blocks */
+	for (unsigned i = 0; i < bbs->bb_cnt; i++) {
+		fprintf(recovery_file_name, "%llu %u\n",
```

Comment:
`ASSERT(bbs->bbv[i].offset != 0 && bbs->bbv[i].length != 0);`

Just to make sure we are not writing out the finish flag early.

Reply:
or rather: ```ASSERT(bbs->bbv[i].length != 0);```
An offset can be zero.

Reply:
Hmm... yup, the length check is enough.